// @geonode/repocket-peer-client/dist/src/shared/settings.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Settings = void 0;
const singleton_1 = require("@ood/singleton");
let Settings = exports.Settings = class Settings {
    constructor() {
        this.settings = {
            is_reusable_socket_mode_on: true,
            is_stand_by_socket_mode_on: true,
            number_of_stand_by_sockets: 2,
        };
        this.convert = i => Boolean(Number(i));
    }
    set({ reusableMode, standByMode, numberOfStandBySockets }) {
        this.settings.is_reusable_socket_mode_on = this.convert(reusableMode);
        this.settings.is_stand_by_socket_mode_on = this.convert(standByMode);
        this.settings.number_of_stand_by_sockets = Number(numberOfStandBySockets);
    }
};
exports.Settings = Settings = __decorate([
    singleton_1.singleton
], Settings);
const globalSettings = new Settings();
exports.default = globalSettings;
